### PR TITLE
Simplify Png Decoding / Encoding

### PR DESCRIPTION
**Pull request by iamcarbon**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

- Makes the PngChunk an immutable strct.
- Changes PngChunkType from a string to an integer

### CODE REVIEW

**Comment by iamcarbon** [comment]
Do we know why this BytesToGeneric test has been randomly failing?

**Comment by iamcarbon** [comment]
Ready for review.

**Comment by JimBobSquarePants** [comment]
@iamcarbon Do we have any before/after benchmarks for this?

**Comment by iamcarbon** [comment]
Allocated 800B less during decoding. Nothing else notable in the benchmarks (within margins of error).

# DecodePng

Before
```
               Method | Runtime |      TestImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |-------- |--------------- |---------:|----------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
 'System.Drawing Png' |    Core | Png/splash.png | 3.012 ms | 3.8100 ms | 0.2153 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 239.95 KB |
     'ImageSharp Png' |    Core | Png/splash.png | 4.793 ms | 1.5015 ms | 0.0848 ms |   1.60 |     0.10 |       - |       - |       - |   3.03 KB |
                      |         |                |          |           |           |        |          |         |         |         |           |
 'System.Drawing Png' |     Clr | Png/splash.png | 2.689 ms | 0.8486 ms | 0.0479 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
     'ImageSharp Png' |     Clr | Png/splash.png | 9.623 ms | 5.0874 ms | 0.2874 ms |   3.58 |     0.10 |       - |       - |       - |  97.09 KB |
```

After
```
               Method | Runtime |      TestImage |     Mean |      Error |    StdDev | Scaled | ScaledSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
--------------------- |-------- |--------------- |---------:|-----------:|----------:|-------:|---------:|--------:|--------:|--------:|----------:|
 'System.Drawing Png' |    Core | Png/splash.png | 3.018 ms |  3.0733 ms | 0.1736 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 239.95 KB |
     'ImageSharp Png' |    Core | Png/splash.png | 4.975 ms |  3.6461 ms | 0.2060 ms |   1.65 |     0.09 |       - |       - |       - |   2.21 KB |
                      |         |                |          |            |           |        |          |         |         |         |           |
 'System.Drawing Png' |     Clr | Png/splash.png | 2.849 ms |  0.4407 ms | 0.0249 ms |   1.00 |     0.00 | 74.2188 | 74.2188 | 74.2188 | 240.54 KB |
     'ImageSharp Png' |     Clr | Png/splash.png | 9.428 ms | 14.0265 ms | 0.7925 ms |   3.31 |     0.23 |       - |       - |       - |  96.34 KB |
```
# EncodePng

Before
```
               Method | Runtime | LargeImage |      Mean |    Error |    StdDev | Scaled | ScaledSD |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
--------------------- |-------- |----------- |----------:|---------:|----------:|-------:|---------:|---------:|---------:|---------:|----------:|
 'System.Drawing Png' |    Core |      False |  22.63 ms | 24.47 ms | 1.3828 ms |   1.00 |     0.00 | 343.7500 | 187.5000 | 187.5000 |   2.03 MB |
     'ImageSharp Png' |    Core |      False |  74.60 ms | 58.82 ms | 3.3233 ms |   3.31 |     0.21 | 187.5000 | 187.5000 | 187.5000 |   1.18 MB |
                      |         |            |           |          |           |        |          |          |          |          |           |
 'System.Drawing Png' |     Clr |      False |  21.69 ms | 25.55 ms | 1.4439 ms |   1.00 |     0.00 | 343.7500 | 187.5000 | 187.5000 |   2.03 MB |
     'ImageSharp Png' |     Clr |      False | 103.67 ms | 15.55 ms | 0.8785 ms |   4.79 |     0.26 | 187.5000 | 187.5000 | 187.5000 |   1.18 MB |
```
After
```
               Method | Runtime | LargeImage |     Mean |     Error |    StdDev | Scaled | ScaledSD |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
--------------------- |-------- |----------- |---------:|----------:|----------:|-------:|---------:|---------:|---------:|---------:|----------:|
 'System.Drawing Png' |    Core |      False | 20.14 ms | 31.761 ms | 1.7946 ms |   1.00 |     0.00 | 343.7500 | 187.5000 | 187.5000 |   2.03 MB |
     'ImageSharp Png' |    Core |      False | 76.76 ms |  9.591 ms | 0.5419 ms |   3.90 |     0.28 | 187.5000 | 187.5000 | 187.5000 |   1.18 MB |
                      |         |            |          |           |           |        |          |          |          |          |           |
 'System.Drawing Png' |     Clr |      False | 22.14 ms | 19.755 ms | 1.1162 ms |   1.00 |     0.00 | 343.7500 | 187.5000 | 187.5000 |   2.03 MB |
     'ImageSharp Png' |     Clr |      False | 95.43 ms | 44.096 ms | 2.4915 ms |   4.32 |     0.20 | 187.5000 | 187.5000 | 187.5000 |   1.18 MB |
```

File: src/ImageSharp/Formats/Png/PngChunkType.cs [hunk 1]
```diff
@@ -4,57 +4,57 @@
...
-    internal static class PngChunkTypes
+    internal enum PngChunkType : uint
     {
         /// <summary>
         /// The first chunk in a png file. Can only exists once. Contains
         /// common information like the width and the height of the image or
         /// the used compression method.
         /// </summary>
-        public const string Header = "IHDR";
+        Header = 1229472850U,       // IHDR
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngChunkType.cs]: I think these would be better expressed in hex so we can match them should we ever require a hex editor.

File: src/ImageSharp/Formats/Png/PngChunk.cs [hunk 2]
```diff
@@ -8,11 +8,14 @@ namespace SixLabors.ImageSharp.Formats.Png
     /// <summary>
     /// Stores header information about a chunk.
     /// </summary>
-    internal sealed class PngChunk
+    internal readonly struct PngChunk
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngChunk.cs]: Much cleaner!

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs [hunk 3]
```diff
@@ -622,33 +616,27 @@ private void WriteChunk(Stream stream, string type, byte[] data)
...
         {
-            BinaryPrimitives.WriteInt32BigEndian(this.intBuffer, length);
-
-            stream.Write(this.intBuffer, 0, 4); // write the length
-
-            this.chunkTypeBuffer[0] = (byte)type[0];
-            this.chunkTypeBuffer[1] = (byte)type[1];
-            this.chunkTypeBuffer[2] = (byte)type[2];
-            this.chunkTypeBuffer[3] = (byte)type[3];
+            BinaryPrimitives.WriteInt32BigEndian(this.buffer, length);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngEncoderCore.cs]: I love seeing code like this replaced with `BinaryPrimitives`

File: tests/ImageSharp.Tests/Formats/Png/PngChunkTypeTests.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+using Xunit;
+
+namespace SixLabors.ImageSharp.Tests.Formats.Png
+{
+    public class PngChunkTypeTests
+    {
+        [Fact]
+        public void ChunkTypeIdsAreCorrect()
+        {
+            Assert.Equal(PngChunkType.Header,       GetType("IHDR"));
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngChunkTypeTests.cs]: 👍 Was wondering about this.

File: tests/ImageSharp.Tests/Formats/Png/PngDecoderTests.cs [hunk 5]
```diff
@@ -242,10 +242,10 @@ public void DetectPixelSize(string imagePath, int expectedPixelSize)
         }
 
         [Theory]
-        [InlineData(PngChunkTypes.Header)]
-        [InlineData(PngChunkTypes.Palette)]
+        [InlineData("IHDR")] // Header
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngDecoderTests.cs]: Could we refactor this test to use the `PngChunkType` enum + `BinaryPrimitives`?

**Pull request merged**